### PR TITLE
Apply fog on sky reflections

### PR DESCRIPTION
**Pull request by adrien-de-tocqueville**
### Purpose of this PR

https://fogbugz.unity3d.com/f/cases/1357622/

Apply fog when sampling sky texture for reflections. Doesn't apply volumetric fog.
Currently, fog is applied in screenspace after rendering, so the sky env texture doesn't contain fog.
SSR correctly reflects fog since it samples the rendered texture, but sampling the cubemap doesn't
So we should apply fog between pixel position and skybox when sampling the sky to have a consistent look between reflections.
But applying volumetric fog requires sampling the vbuffer which is kind of expensive

Following screenshot have a black plane with smoothness = 1
SSR is enabled, which is why some part of the reflection looks correct

|   | Master | Branch |
| ------------- | ------------- | ------------- |
| Fog only  | image  | image |
| Fog + Volumetric Fog  | image  | image |

### Testing status

Tested on the repro project

Tested fog color, max distance, height worked correctly

I had to update the screenshots of a test because it was using fog, so reflections were not correct

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoopDef.hlsl [hunk 1]
```diff
@@ -18,6 +18,7 @@ struct HDShadowContext { float unused; };
 struct LightLoopContext
 {
     int sampleReflection;
+    float3 positionWS;
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoopDef.hlsl:21]: add a comment about why we need it here.  could be under ifdef #ifdef UNITY_ATMOSPHERIC_SCATTERING_INCLUDED

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoopDef.hlsl [hunk 2]
```diff
@@ -164,6 +165,21 @@ float4 SampleEnv(LightLoopContext lightLoopContext, int index, float3 texCoord,
         color.rgb = SampleSkyTexture(texCoord, lod, sliceIdx).rgb;
         // Sky isn't pre-expose, so best to clamp to max16 here in case of inf
         color.rgb = ClampToFloat16Max(color.rgb);
+
+#ifdef UNITY_ATMOSPHERIC_SCATTERING_INCLUDED
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoopDef.hlsl:169]: this is currently not define anywhere

**Review by sebastienlagarde — COMMENTED** [review]
appraoch looks good

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoopDef.hlsl (hunk 2 above)
Comment (adrien-de-tocqueville) [reply com.unity.render-pipelines.high-definition/Runtime/Lighting/LightLoop/LightLoopDef.hlsl:169]: it's defined when the atmospheric scattering file is included, not very clean but it works at least for a draft

**Comment by adrien-de-tocqueville** [comment]
Note: when a material has clear coat, fog attenuation is computed a second time for the clear coat layer
Although it's correct, ut's maybe not necessary (because both attenuation should be very similar)
But it would overcomplicate the code or change the shader api for a small performance gain in a rare case

**Review by JMargevics — APPROVED** [review]
Screen Shot 2022-02-18 at 13 15 19
Verified on metal player. Volumetric fog mixes nicely with default sky probe now  👍🟢

**Comment by JMargevics** [comment]
One thing: should we cover this somewhere in automation tests?

**Comment by sebastienlagarde** [comment]
@adrien-de-tocqueville , see with @JulienIgnace-Unity  if you can land this fix

**Review by JulienIgnace-Unity — APPROVED** [review]

**Comment by adrien-de-tocqueville** [comment]
@JulienIgnace-Unity i think the PR is ready, looking at failing test on yamato it seems they are the expected ones (but i am not a professional)

**Comment by adrien-de-tocqueville** [comment]
will reopen on ono

**Pull request closed without merging**